chain/jsonrpc-primitives/src/types/validator.rs

```diff
@@ -0,0 +1,89 @@
+use near_primitives::types::EpochReference;
+use near_primitives::views::EpochValidatorInfo;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+#[derive(thiserror::Error, Debug)]
+pub enum RpcValidatorError {
+    #[error("Epoch not found")]
+    UnknownEpoch,
+    #[error("The node reached its limits. Try again later. More details: {0}")]
+    InternalError(String),
+    // NOTE: Currently, the underlying errors are too broad, and while we tried to handle
+    // expected cases, we cannot statically guarantee that no other errors will be returned
+    // in the future.
+    // TODO #3851: Remove this variant once we can exhaustively match all the underlying errors
+    #[error("It is a bug if you receive this error type, please, report this incident: https://github.com/near/nearcore/issues/new/choose. Details: {0}")]
+    Unreachable(String),
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RpcValidatorRequest {
+    #[serde(flatten)]
+    pub epoch_reference: EpochReference,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RpcValidatorResponse {
+    #[serde(flatten)]
+    pub validator_info: EpochValidatorInfo,
+}
+
+impl From<near_client_primitives::types::GetValidatorInfoError> for RpcValidatorError {
+    fn from(error: near_client_primitives::types::GetValidatorInfoError) -> Self {
+        match error {
+            near_client_primitives::types::GetValidatorInfoError::UnknownEpoch => {
+                Self::UnknownEpoch
+            }
+            near_client_primitives::types::GetValidatorInfoError::IOError(s) => {
+                Self::InternalError(s)
+            }
+            near_client_primitives::types::GetValidatorInfoError::Unreachable(s) => {
+                near_metrics::inc_counter_vec(
+                    &crate::metrics::RPC_UNREACHABLE_ERROR_COUNT,
+                    &["RpcValidatorError", &s],
+                );
+                Self::Unreachable(s)
+            }
+        }
+    }
+}
+
+impl From<actix::MailboxError> for RpcValidatorError {
+    fn from(error: actix::MailboxError) -> Self {
+        Self::InternalError(error.to_string())
+    }
+}
+
+impl RpcValidatorRequest {
+    pub fn parse(
+        value: Option<Value>,
+    ) -> Result<RpcValidatorRequest, crate::errors::RpcParseError> {
+        let epoch_reference = if let Ok((block_id,)) =
+            crate::utils::parse_params::<(near_primitives::types::MaybeBlockId,)>(value.clone())
+        {
+            if block_id.is_some() {
+                return Err(crate::errors::RpcParseError(
+                    "Require at least one parameter".to_owned(),
+                ));
```

Comment:
Can we still let people specify block hash/block height and resolve it to the epoch id? I mean, we won't provide the information as of the arbitrary block hash/height, but rather round it to the epoch boundary instead.